// @tarojs/mobx-h5/dist/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var taroH5 = require('@tarojs/taro-h5');
var nervjs = require('nervjs');
var mobx = require('mobx');

function _typeof(obj) {
  if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") {
    _typeof = function (obj) {
      return typeof obj;
    };
  } else {
    _typeof = function (obj) {
      return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj;
    };
  }

  return _typeof(obj);
}

function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
}

function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ("value" in descriptor) descriptor.writable = true;
    Object.defineProperty(target, descriptor.key, descriptor);
  }
}

function _createClass(Constructor, protoProps, staticProps) {
  if (protoProps) _defineProperties(Constructor.prototype, protoProps);
  if (staticProps) _defineProperties(Constructor, staticProps);
  return Constructor;
}

function _inherits(subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function");
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      writable: true,
      configurable: true
    }
  });
  if (superClass) _setPrototypeOf(subClass, superClass);
}

function _getPrototypeOf(o) {
  _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) {
    return o.__proto__ || Object.getPrototypeOf(o);
  };
  return _getPrototypeOf(o);
}

function _setPrototypeOf(o, p) {
  _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) {
    o.__proto__ = p;
    return o;
  };

  return _setPrototypeOf(o, p);
}

function _assertThisInitialized(self) {
  if (self === void 0) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return self;
}

function _possibleConstructorReturn(self, call) {
  if (call && (typeof call === "object" || typeof call === "function")) {
    return call;
  }

  return _assertThisInitialized(self);
}

function _classCallCheck$1(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
}

function _defineProperties$1(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ("value" in descriptor) descriptor.writable = true;
    Object.defineProperty(target, descriptor.key, descriptor);
  }
}

function _createClass$1(Constructor, protoProps, staticProps) {
  if (protoProps) _defineProperties$1(Constructor.prototype, protoProps);
  if (staticProps) _defineProperties$1(Constructor, staticProps);
  return Constructor;
}

function _defineProperty$1(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

function _inherits$1(subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function");
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      writable: true,
      configurable: true
    }
  });
  if (superClass) _setPrototypeOf$1(subClass, superClass);
}

function _getPrototypeOf$1(o) {
  _getPrototypeOf$1 = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf$$1(o) {
    return o.__proto__ || Object.getPrototypeOf(o);
  };
  return _getPrototypeOf$1(o);
}

function _setPrototypeOf$1(o, p) {
  _setPrototypeOf$1 = Object.setPrototypeOf || function _setPrototypeOf$$1(o, p) {
    o.__proto__ = p;
    return o;
  };

  return _setPrototypeOf$1(o, p);
}

function _assertThisInitialized$1(self) {
  if (self === void 0) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return self;
}

function _possibleConstructorReturn$1(self, call) {
  if (call && (_typeof(call) === "object" || typeof call === "function")) {
    return call;
  }

  return _assertThisInitialized$1(self);
}

function _superPropBase$1(object, property) {
  while (!Object.prototype.hasOwnProperty.call(object, property)) {
    object = _getPrototypeOf$1(object);
    if (object === null) break;
  }

  return object;
}

function _get$1(target, property, receiver) {
  if (typeof Reflect !== "undefined" && Reflect.get) {
    _get$1 = Reflect.get;
  } else {
    _get$1 = function _get$$1(target, property, receiver) {
      var base = _superPropBase$1(target, property);

      if (!base) return;
      var desc = Object.getOwnPropertyDescriptor(base, property);

      if (desc.get) {
        return desc.get.call(receiver);
      }

      return desc.value;
    };
  }

  return _get$1(target, property, receiver || target);
}

function isStateless(component) {
  return !(component.prototype && component.prototype.render);
}

function observer(Component) {
  if (typeof Component !== 'function' || isStateless(Component)) {
    throw new Error("Please pass a valid component to 'observer'");
  }

  if (Component.isMobxInjector === true) {
    console.warn("Mobx observer: You are trying to use 'observer' on a component that already has 'inject'. Please apply 'observer' before applying 'inject'");
  }

  var ObserverComponent =
  /*#__PURE__*/
  function (_Component) {
    _inherits$1(ObserverComponent, _Component);

    function ObserverComponent() {
      _classCallCheck$1(this, ObserverComponent);

      return _possibleConstructorReturn$1(this, _getPrototypeOf$1(ObserverComponent).apply(this, arguments));
    }

    _createClass$1(ObserverComponent, [{
      key: "componentWillMount",
      value: function componentWillMount() {
        var _this = this;

        var initialName = this.displayName || this.name || this.constructor && (this.constructor.displayName || this.constructor.name) || '<component>';
        this._reaction = new mobx.Reaction("".concat(initialName, "_").concat(Date.now()), function () {
          if (typeof _this.componentWillReact === 'function') {
            _this.componentWillReact();
          }

          _this.forceUpdate();
        });

        if (typeof _get$1(_getPrototypeOf$1(ObserverComponent.prototype), "componentWillMount", this) === 'function') {
          _get$1(_getPrototypeOf$1(ObserverComponent.prototype), "componentWillMount", this).call(this);
        }
      }
    }, {
      key: "componentWillUnmount",
      value: function componentWillUnmount() {
        this._reaction.dispose();

        if (typeof _get$1(_getPrototypeOf$1(ObserverComponent.prototype), "componentWillUnmount", this) === 'function') {
          _get$1(_getPrototypeOf$1(ObserverComponent.prototype), "componentWillUnmount", this).call(this);
        }
      }
    }]);

    return ObserverComponent;
  }(Component);

  var target = ObserverComponent.prototype;
  var originRender = target.render;

  target.render = function () {
    var _this2 = this;

    var result;
    var exception;

    if (this._reaction && this._reaction instanceof mobx.Reaction) {
      this._reaction.track(function () {
        try {
          result = mobx._allowStateChanges(false, originRender.bind(_this2));
        } catch (e) {
          exception = e;
        }
      });
    } else {
      result = originRender.call(this);
    }

    if (exception) {
      throw exception;
    }

    return result;
  };

  return ObserverComponent;
}

var store = {};

function getStore() {
  return store;
}

function setStore(arg) {
  store = arg;
}

var proxiedInjectorProps = {
  isMobxInjector: {
    value: true,
    writable: true,
    configurable: true,
    enumerable: true
    /**
     * Store Injection
     */

  }
};

function createStoreInjector(grabStoresFn, component, injectNames, _ref) {
  var Component = _ref.Component,
      createElement = _ref.createElement;
  var displayName = 'inject-' + (component.displayName || component.name || component.constructor && component.constructor.name || 'Unknown');
  if (injectNames) displayName += '-with-' + injectNames;

  var Injector =
  /*#__PURE__*/
  function (_Component) {
    _inherits$1(Injector, _Component);

    function Injector() {
      _classCallCheck$1(this, Injector);

      return _possibleConstructorReturn$1(this, _getPrototypeOf$1(Injector).apply(this, arguments));
    }

    _createClass$1(Injector, [{
      key: "render",
      value: function render() {
        var newProps = {};

        for (var key in this.props) {
          if (this.props.hasOwnProperty(key)) {
            newProps[key] = this.props[key];
          }
        }

        var additionalProps = grabStoresFn(getStore() || {}, newProps, this.context) || {};

        for (var _key in additionalProps) {
          newProps[_key] = additionalProps[_key];
        }

        return createElement(component, newProps);
      }
    }]);

    return Injector;
  }(Component);

  _defineProperty$1(Injector, "displayName", displayName);

  Object.defineProperties(Injector, proxiedInjectorProps);
  return Injector;
}

function grabStoresByName(storeNames) {
  return function (baseStores, nextProps) {
    storeNames.forEach(function (storeName) {
      if (!(storeName in baseStores)) {
        throw new Error("MobX injector: Store '" + storeName + "' is not available! Make sure it is provided by some Provider");
      }

      nextProps[storeName] = baseStores[storeName];
    });
    return nextProps;
  };
}

function inject()
/* fn(stores, nextProps) or ...storeNames, { Component, createElement } */
{
  var grabStoresFn;
  var platformSpecialParams = arguments[arguments.length - 1];

  if (typeof arguments[0] === 'function') {
    grabStoresFn = arguments[0];
    return function (componentClass) {
      return createStoreInjector(grabStoresFn, componentClass, null, platformSpecialParams);
    };
  } else {
    var storeNames = [];

    for (var i = 0; i < arguments.length - 1; i++) {
      storeNames[i] = arguments[i];
    }

    grabStoresFn = grabStoresByName(storeNames);
    return function (componentClass) {
      return createStoreInjector(grabStoresFn, componentClass, storeNames.join('-'), platformSpecialParams);
    };
  }
}

var Provider =
/*#__PURE__*/
function (_Component) {
  _inherits(Provider, _Component);

  function Provider(props) {
    var _this;

    _classCallCheck(this, Provider);

    _this = _possibleConstructorReturn(this, _getPrototypeOf(Provider).call(this, props));
    setStore(props.store);
    return _this;
  }

  _createClass(Provider, [{
    key: "render",
    value: function render() {
      return nervjs.Children.only(this.props.children);
    }
  }]);

  return Provider;
}(taroH5.Component);

function inject$1() {
  return inject.apply(void 0, Array.prototype.slice.call(arguments).concat([{
    Component: taroH5.Component,
    createElement: nervjs.createElement
  }]));
}

exports.inject = inject$1;
exports.observer = observer;
exports.Provider = Provider;
